lib/ansible/executor/task_executor.py

```diff
@@ -1106,8 +1116,16 @@ def _get_action_handler(self, connection, templar):
 
         collections = self._task.collections
 
+        # Check if the module has specified to action handler
+        module = None
+        if self._shared_loader_obj.module_loader.has_plugin(self._task.action, collection_list=collections):
```

Comment:
This change is entirely to support the new module routing keyword. Since `_load_module_defaults` now does not merge all module defaults to the same action plugin name, TE needs to access the defaults using the same name the defaults were stored under and ensure the action_plugin specified by the module gets used. It shouldn't be a breaking change unless someone decided to add a action_plugin entry for module routing before it did anything. I could tighten this up a little to set the module context back to None if it isn't resolved or defined the action plugin it needs:
```diff
diff --git a/lib/ansible/executor/task_executor.py b/lib/ansible/executor/task_executor.py
index 6e699285b0..2c51f8ee32 100644
--- a/lib/ansible/executor/task_executor.py
+++ b/lib/ansible/executor/task_executor.py
@@ -1122,7 +1122,9 @@ class TaskExecutor:
             module = self._shared_loader_obj.module_loader.find_plugin_with_context(
                 self._task.action, collection_list=self._task.collections
             )
-        if module is not None and module.resolved and module.action_plugin:
+        if module is None or not module.resolved or not module.action_plugin:
+            module = None
+        if module is not None:
```
I'll try to consolidate this to just one call as well - I can't remember why I did this, perhaps there's an error if the module does not exist (in which case it falls back to existing behavior).

Reply:
it really does not break backwards compat unless people were using the new key already